pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+			Self::Rate,
+			Self::Balance,
+			Adjustment<Self::Balance>,
+			NormalizedDebt = Self::Balance,
+		>;
+
+		/// Max number of active loans per pool.
+		#[pallet::constant]
+		type MaxActiveLoansPerPool: Get<u32>;
+
+		/// Max number of write-off groups per pool.
+		#[pallet::constant]
+		type MaxWriteOffGroups: Get<u32>;
+	}
+
+	/// Contains the last loan id generated
+	#[pallet::storage]
+	pub(crate) type LastLoanId<T: Config> = StorageValue<_, T::LoanId, ValueQuery, GetDefault>;
+
+	/// Storage for loans that has been created but are not still active.
+	#[pallet::storage]
+	pub(crate) type CreatedLoans<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		PoolIdOf<T>,
+		Blake2_128Concat,
+		T::LoanId,
+		CreatedLoan<T>,
+		OptionQuery,
+	>;
+
+	/// Storage for active loans.
+	/// The indexation of this storage changes regarding the `CreatedLoans` or `ClosedLoans`
+	/// because here we try to minimize the iteration speed over all active loans in a pool.
+	#[pallet::storage]
+	pub(crate) type ActiveLoans<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		PoolIdOf<T>,
+		BoundedVec<ActiveLoan<T>, T::MaxActiveLoansPerPool>,
+		ValueQuery,
+	>;
+
+	/// Storage for closed loans.
+	/// No mutations are expected in this storage.
+	/// Loans are stored here for historical purposes.
+	#[pallet::storage]
+	pub(crate) type ClosedLoans<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		PoolIdOf<T>,
+		Blake2_128Concat,
+		T::LoanId,
+		ClosedLoan<T>,
+		OptionQuery,
+	>;
+
+	/// Stores write off policy used in each pool
+	#[pallet::storage]
+	pub(crate) type WriteOffPolicy<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		PoolIdOf<T>,
+		BoundedVec<WriteOffState<T::Rate>, T::MaxWriteOffGroups>,
+		ValueQuery,
+	>;
+
+	/// Stores the portfolio valuation associated to each pool
+	#[pallet::storage]
+	pub(crate) type LatestPortfolioValuations<T: Config> =
+		StorageMap<_, Blake2_128Concat, PoolIdOf<T>, PortfolioValuation<T::Balance>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A loan was created
+		Created {
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			loan_info: LoanInfo<T>,
+		},
+		/// An amount was borrowed for a loan
+		Borrowed {
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			amount: T::Balance,
+		},
+		/// An amount was repaid for a loan
+		Repaid {
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			amount: T::Balance,
+		},
+		/// A loan was written off
+		WrittenOff {
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			status: WriteOffStatus<T::Rate>,
+		},
+		/// A loan was closed
+		Closed {
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			collateral: AssetOf<T>,
+		},
+		/// The Portfolio Valuation for a pool was updated.
+		PortfolioValuationUpdated {
+			pool_id: PoolIdOf<T>,
+			value: T::Balance,
+			update_type: PortfolioValuationUpdateType,
+		},
+		WriteOffPoliciesUpdated {
```

Comment:
A small renaming, but also, I think this event can include the new policy?
```suggestion
		WriteOffPolicyUpdated {
```

Reply:
Sure!